p2p/protocol/head/head.go

```diff
@@ -0,0 +1,99 @@
+package head
+
+import (
+	"context"
+	"encoding/json"
+	"net"
+	"net/http"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p-core/host"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	gostream "github.com/libp2p/go-libp2p-gostream"
+)
+
+const closeTimeout = 30 * time.Second
+
+var log = logging.Logger("go-legs-head")
+
+func deriveProtocolID(topic string) protocol.ID {
+	return protocol.ID("/legs/head/" + topic + "/0.0.1")
+}
+
+func (p *Publisher) Serve(host host.Host, topic string) error {
+	l, err := gostream.Listen(host, deriveProtocolID(topic))
+	if err != nil {
+		return err
+	}
+
+	p.server = &http.Server{Handler: http.Handler(p)}
+	return p.server.Serve(l)
+}
+
+func QueryRootCid(ctx context.Context, host host.Host, topic string, peer peer.ID) (cid.Cid, error) {
+	client := http.Client{
+		Transport: &http.Transport{
+			DialContext: func(ctx context.Context, network, addr string) (net.Conn, error) {
+				return gostream.Dial(ctx, host, peer, deriveProtocolID(topic))
+			},
+		},
+	}
+
+	// The httpclient expects there to be a host here. `.invalid` is a reserved
+	// TLD for this purpose. See
+	// https://datatracker.ietf.org/doc/html/rfc2606#section-2
+	resp, err := client.Get("http://unused.invalid/head")
+	if err != nil {
+		return cid.Undef, err
+	}
+	defer resp.Body.Close()
+
+	var cidStr string
+	err = json.NewDecoder(resp.Body).Decode(&cidStr)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	return cid.Decode(cidStr)
+}
+
+type Publisher struct {
+	rl     sync.RWMutex
+	root   cid.Cid
+	server *http.Server
+}
+
+func (p *Publisher) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if path.Base(r.URL.Path) != "head" {
+		http.Error(w, "", http.StatusNotFound)
+		return
+	}
+
+	p.rl.RLock()
+	defer p.rl.RUnlock()
+	out, err := json.Marshal(p.root.String())
+	if err != nil {
+		log.Infow("failed to serve root", "err", err)
+		http.Error(w, "", http.StatusInternalServerError)
+	} else {
```

Comment:
nit: return after `http.Error` and no else... since there is nothing to do after `http.Error`, so less chance of accidentally not returning.